### ISSUE TITLE
[Proposal] `Assert` function

### ISSUE THREAD

**Issue by BowenBao** [microsoft/onnx-script#500]
The `Assert` function, together with integration of the evaluator, provides runtime check for `type`, `shape`, `value` or `attributes`. The main goal of this function is to
1. Catch unexpected scenarios and provide verbose explanations explaining the failure. 
   Examples:
    * Composing an atenlib operator that does not support dynamic input shapes or certain types.
    * Exporter to convert upstream fx guards as `Assert`.
2. Guard against model code modification that breaks pre-existing assumptions.
    Examples:
    * Capturing `trace_only=True` as local function, and guard that node inputs/attributes are not hand modified later to break the conditions that were statically traced.

This function should be equivalent to a no-op by default if not supported by backend.

Initially I thought to propose `Assert` as new operator to ONNX. Yet I feel onnx-script might be a better place to start, simply because
it needs backend integration for this op to be useful. Hence it is much faster to validate the idea and iterate on onnx-script and its own eager evaluator.

Written off quickly per discussion and as follow up to #443 as brainstorm idea. This is nowhere near a complete proposal. Feedbacks and discussions are welcomed.

cc @titaiwangms @justinchuby @gramalingam @xadupre

**Comment by justinchuby** [comment]
Curious to learn more about what it looks like when used?

**Comment by BowenBao** [comment]
@justinchuby added an example case for illustration.

**Comment by justinchuby** [comment]
I see what it's doing! We can may be able to call it `assert_equal` and return a string as the message?

OK if passed, Failed or msg if failed.

**Comment by xadupre** [comment]
Node not connected to the outputs are usually ignored by runtimes. How do you plan to use the output? I guess this operator would only be used to check the shape or the type of a tensor (otherwise, the model size would increase a lot). Then, it is really needed only when the shape inference fails to guess the shape. I wonder if there is a way to leverage this new information during shape inference, which could be useful for every domain except ai.onnx or ai.onnx.ml.

**Comment by justinchuby** [comment]
I wonder if we can set it as output when we debug.

**Comment by gramalingam** [comment]
An "assert" op in ONNX would make sense.

But can we get away with a single primitive (in ONNX) that takes a boolean (to be checked) and a string (the error-message) as arguments?

I think we can define things like `assert_equal` as functions (using onnxscript) over this single primitive.

**Comment by justinchuby** [comment]
Do we want runtime failures with Assert?

**Comment by gramalingam** [comment]
> Do we want runtime failures with Assert?

That seems the simplest semantics.

**Comment by justinchuby** [comment]
Sounds good to me!

**Comment by BowenBao** [comment]
> Node not connected to the outputs are usually ignored by runtimes. How do you plan to use the output?

My thought was that `assert` would require special backend integration anyway, to be able to raise/throw, hence it may be specialized to not be ignored when not connected. For backends that don't support it (have no knowledge of it), being ignored is probably good.

> I guess this operator would only be used to check the shape or the type of a tensor (otherwise, the model size would increase a lot). Then, it is really needed only when the shape inference fails to guess the shape. I wonder if there is a way to leverage this new information during shape inference, which could be useful for every domain except ai.onnx or ai.onnx.ml.

It's good point about model size. I think still it might be useful to check on some small constants (e.g., axes values). Also It can check attribute values, when used inside another function.

**Comment by justinchuby** [comment]
I understand the data contributes to most of the model size right? So a few more nodes wouldn’t be too bad

**Comment by gramalingam** [comment]
> I understand the data contributes to most of the model size right? So a few more nodes wouldn’t be too bad

Yes. I think Xavier's point was about checking equality of tensors by specifying expected-value, which could be large.

**Comment by BowenBao** [comment]
Yea technically this does allow stuffing in arbitrary constant tensors as check and blow up the model size. Since exporter and atenlib(onnxscript) are the two potential main users, I'm not too worried about this getting abused. That being said we can always warn/error against this, like complain in checker.
